rfd/0209-mcp-access.md

```diff
@@ -0,0 +1,359 @@
+---
+authors: STeve Huang (xin.huang@goteleport.com)
+state: draft
+---
+
+# RFD 0209 - MCP Access
+
+## Required Approvers
+
+* Engineering: @r0mant
+* Product: @klizhentas
+
+## What
+
+Support zero-trust access for MCP servers.
+
+## Why
+
+Introduced in late 2024, Anthropic’s [Model Context Protocol
+(MCP)](https://modelcontextprotocol.io/introduction) is a widely adopted,
+open-source standard that enables language models to seamlessly interact with
+external tools and data, enhancing their contextual capabilities.
+
+However, MCP servers today are mostly operated locally without secure transport
+or authorization. OAuth authorization was recently added to the specification,
+but as of the time of writing, it is still new and not widely adopted.
+
+With Teleport's support for MCP servers, users can:
+- Access MCP servers on remote machines where transport is secured with TLS
+- Control access to specific MCP servers and specific tools defined by Teleport roles
+- Track activity through Teleport's audit log
+
+## Details
+
+To expedite the initial rollout, the first iteration of MCP server support will
+be implemented through App access.
+
+### UX
+
+#### editor - configure a filesystem MCP server and define RBAC
+To configure a [filesystem MCP
+server](https://github.com/modelcontextprotocol/servers/tree/main/src/filesystem)
+via Teleport app service:
+```yaml
+app_service:
+  enabled: true
+  apps:
+  - name: "dev-files"
+    description: "Shared files for developers"
+    labels:
+      env: "dev"
+    mcp:
+      command: "npx"
+      args: ["-y", "@modelcontextprotocol/server-filesystem", "/home/dev/files"]
+```
+
+To create a role for admins that have full access to MCP servers:
+```yaml
+kind: role
+version: v7
+metadata:
+  name: admin
+spec:
+  allow:
+    app_labels:
+      "*": "*"
+```
+
+To create a role for devs that can only access `dev` MCP servers and only have
+read-only access to the filesystem:
+```yaml
+kind: role
+version: v7
+metadata:
+  name: dev
+spec:
+  allow:
+    app_labels:
+      "env": "dev"
+    # The name of the MCP tools to allow access to.
+    # The wildcard character '*' matches any sequence of characters.
+    # If the value begins with '^' and ends with '$', it is treated as a regular expression.
+    mcp_tools:
+    - search_files
+    - get_*
+    - ^(read|list)_.*$
+```
+
+#### access - configure Claude Desktop to use the MCP server via Teleport
+
+First, to retrieve a list of allowed MCP servers:
+```bash
+$ tsh login ...
+...
+
+$ tsh mcp ls
+Name      Description                 Command Args
+--------- --------------------------- ------- ------------------------------------------------
+dev-files Shared files for developers npx     [-y @modelcontextprotocol/server-filesystem ...]
```

Comment:
I wonder how really useful is the "Command" and "Args" command to average end user. Maybe hide it be default unless specific flag is passed?

Reply:
+1 on this, but the ideal scenario would be to present a short overview of the tools/resources that the MCP offers, so users don't need to hop into an MCP client to see what's available. As for the initial version, showing the command would help because users could search the MCP and see its definition.

Reply:
Moved `Command` and `args` to `-v`, and added `Type` and `Labels` for `tsh mcp ls` without verbose.

Reply:
Thinking about this further, we could inspect declared tools and resources that server offers and show that info here. Just tool names might be useful, with an option for full description under `--format=json`.

All of this post-MVP, though.